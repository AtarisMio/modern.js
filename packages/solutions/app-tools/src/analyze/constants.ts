export const JS_EXTENSIONS = [
  '.js',
  '.jsx',
  '.ts',
  '.tsx',
  '.mjs',
  '.mjsx',
  '.cjs',
  '.cjsx',
];

export const INDEX_FILE_NAME = 'index';

export const APP_FILE_NAME = 'App';

export const PAGES_DIR_NAME = 'pages';

export const ROUTES_DIR_NAME = 'routes';

export const FILE_SYSTEM_ROUTES_FILE_NAME = 'routes.js';

export const LOADERS_FILE_TS = 'loaders.ts';

export const LOADERS_FILE_JS = 'loaders.js';

export const ENTRY_POINT_FILE_NAME = 'index.js';
export const FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP = /^\[(\S+)\]([*+?]?)$/;

export const FILE_SYSTEM_ROUTES_LAYOUT = '_layout';

export const FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT = '_app';

export const FILE_SYSTEM_ROUTES_INDEX = 'index';

export const FILE_SYSTEM_ROUTES_IGNORED_REGEX =
  /\.(d|test|spec|e2e)\.(js|jsx|ts|tsx)$/;

export const HTML_PARTIALS_FOLDER = 'html';

export const HTML_PARTIALS_EXTENSIONS = ['.htm', '.html', '.ejs'];

export const FILE_SYSTEM_ROUTES_COMPONENTS_DIR = 'internal_components';
