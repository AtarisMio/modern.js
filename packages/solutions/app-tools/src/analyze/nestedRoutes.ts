import path from 'path';
import { fs } from '@modern-js/utils';
import { parseModule } from '../utils/parseModule';
import { JS_EXTENSIONS } from './constants';

export const CLIENT_LOADER = 'clientLoader';

export interface NestedRoute {
  id: string;
  parentId?: string;
  path: string;
  componentPath: string;
  index: boolean;
  root: boolean;
  [CLIENT_LOADER]: boolean;
  children: NestedRoute[];
  readonly childIds: string[];
}

const hasClientLoader = async (filename: string) => {
  const source = await fs.readFile(filename);
  const [, moduleExports] = await parseModule({
    source: source.toString(),
    filename,
  });
  const clientLoader = moduleExports.find(e => e.n === CLIENT_LOADER);
  return Boolean(clientLoader);
};

const getPathWithoutExt = (filename: string) => {
  const extname = path.extname(filename);
  return filename.slice(0, -extname.length);
};

const createRoute = (
  routeInfo: Omit<NestedRoute, 'id' | 'children' | 'childIds'>,
  routesDir: string,
): NestedRoute => {
  const { componentPath } = routeInfo;
  const relativePath = path.relative(routesDir, componentPath);
  const id = getPathWithoutExt(relativePath);
  return {
    ...routeInfo,
    id,
    children: [],
    get childIds() {
      return this.children?.map(route => route.id);
    },
  };
};

const createIndexRoute = (
  routeInfo: Omit<
    NestedRoute,
    'id' | 'index' | 'root' | 'children' | 'childIds'
  >,
  routesDir: string,
) => {
  return createRoute(
    {
      ...routeInfo,
      index: true,
      root: false,
    },
    routesDir,
  );
};

const createRootRoute = (
  routeInfo: Omit<
    NestedRoute,
    'id' | 'index' | 'root' | 'children' | 'childIds'
  >,
  routesDir: string,
) => {
  return createRoute(
    {
      ...routeInfo,
      index: false,
      root: true,
    },
    routesDir,
  );
};

const walk = async (
  dirname: string,
  routesDir: string,
): Promise<NestedRoute[]> => {
  if (!(await fs.pathExists(dirname))) {
    return [];
  }

  const isDirectory = (await fs.stat(dirname)).isDirectory();
  if (!isDirectory) {
    return [];
  }

  const routes: NestedRoute[] = [];
  const routeMap = new Map<string, NestedRoute>();
  const routesMap = new Map<string, NestedRoute[]>();
  const isRoot = dirname === routesDir;
  const items = await fs.readdir(dirname);

  let appRoute: NestedRoute | null = null;
  for (const item of items) {
    const itemPath = path.join(dirname, item);
    const extname = path.extname(item);
    const itemWithoutExt = item.slice(0, -extname.length);
    const itemPathWithoutExt = itemPath.slice(0, -extname.length);

    if (
      itemPath.startsWith('_') ||
      (extname && !JS_EXTENSIONS.includes(extname))
    ) {
      continue;
    }

    let route: NestedRoute;
    const isDirectory = (await fs.stat(itemPath)).isDirectory();
    if (isDirectory) {
      const itemRoutes = await walk(itemPath, routesDir);
      const parentRoute = routeMap.get(itemPath);
      if (parentRoute) {
        const { children, childIds } = parentRoute;
        for (const route of itemRoutes) {
          route.parentId = parentRoute.id;
          children.push(route);
          childIds.push(route.id);
        }
        children?.push(...itemRoutes);
        childIds?.push(...itemRoutes.map(route => route.id));
      } else {
        routesMap.set(itemPath, itemRoutes);
      }
      continue;
    }

    if (isRoot && itemWithoutExt === 'app') {
      const clientLoader = await hasClientLoader(itemPath);
      appRoute = createRootRoute(
        {
          path: '/',
          componentPath: itemPath,
          clientLoader,
        },
        routesDir,
      );
      routes.push(appRoute);
      continue;
    }

    if (itemWithoutExt === 'index') {
      const clientLoader = await hasClientLoader(itemPath);
      route = createIndexRoute(
        {
          path: itemWithoutExt,
          componentPath: itemPath,
          clientLoader,
        },
        routesDir,
      );
      routes.push(route);
      if (appRoute) {
        appRoute.children?.push(route);
      }
      continue;
    }

    let finalRoutePath = itemWithoutExt;
    if (/^\[.*\]$/.test(itemWithoutExt)) {
      finalRoutePath = `:${itemWithoutExt.slice(1, itemWithoutExt.length - 1)}`;
    }

    const clientLoader = await hasClientLoader(itemPath);
    route = createRoute(
      {
        path: finalRoutePath,
        componentPath: itemPath,
        index: false,
        root: false,
        clientLoader,
      },
      routesDir,
    );
    routes.push(route);
    if (appRoute) {
      appRoute.children?.push(route);
    }

    const childRoutes = routesMap.get(itemPathWithoutExt);
    if (childRoutes) {
      route.children = childRoutes;
    } else {
      routeMap.set(itemPathWithoutExt, route);
    }
  }

  return routes;
};

// TODO: 是否允许用户不写 app.tsx 呢？@yimingjfe
export const getNestedRoutes = async (routesDir: string) => {
  const routes = await walk(routesDir, routesDir);
  const rootRoute = routes.find(route => route.root);
  return rootRoute;
};
