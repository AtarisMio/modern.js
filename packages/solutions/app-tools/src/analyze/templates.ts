import type { RuntimePlugin } from '@modern-js/core';
import type { Entrypoint, Route } from '@modern-js/types';
import { CLIENT_LOADER, NestedRoute } from './nestedRoutes';

export const index = ({
  mountId,
  imports,
  renderFunction,
  exportStatement,
}: {
  mountId: string;
  imports: string;
  exportStatement: string;
  renderFunction: string;
}) => `
const IS_BROWSER = typeof window !== 'undefined' && window.name !== 'nodejs';
const MOUNT_ID = '${mountId}';

${imports}

let AppWrapper = null;

function render() {
  ${renderFunction}
}

AppWrapper = render();

${exportStatement};
`;

export const renderFunction = ({
  plugins,
  customBootstrap,
  fileSystemRoutes,
}: {
  plugins: RuntimePlugin[];
  customBootstrap?: string | false;
  fileSystemRoutes: Entrypoint['fileSystemRoutes'];
}) => `
  AppWrapper = createApp({
    plugins: [
     ${plugins
       .map(
         ({ name, options, args }) =>
           `${name}({...${options}, ...App?.config?.${args || name}}),`,
       )
       .join('\n')}
    ]
  })(${fileSystemRoutes ? '' : `App`})

  if (IS_BROWSER) {
    ${
      customBootstrap
        ? `customBootstrap(AppWrapper);`
        : `bootstrap(AppWrapper, MOUNT_ID);`
    }
  }

  return AppWrapper
`;

export const html = (partials: {
  top: string[];
  head: string[];
  body: string[];
}) => `
<!DOCTYPE html>
<html>
<head>
  <%= meta %>
  <title><%= title %></title>

  ${partials.top.join('\n')}

  <script>
    window.__assetPrefix__ = '<%= assetPrefix %>';
  </script>
  ${partials.head.join('\n')}

  <!--<?- chunksMap.css ?>-->
</head>

<body>
  <noscript>
    We're sorry but react app doesn't work properly without JavaScript enabled. Please enable it to continue.
  </noscript>
  <div id="<%= mountId %>"><!--<?- html ?>--></div>
  ${partials.body.join('\n')}
  <!--<?- chunksMap.js ?>-->
  <!--<?- SSRDataScript ?>-->
  <!--<?- bottomTemplate ?>-->
</body>

</html>
`;

const flattenRoute = (route: NestedRoute): NestedRoute[] => [
  route,
  ...route.children.flatMap(flattenRoute),
];

export const getClientLoaderCode = (nestedRoute: NestedRoute) => {
  const routes = flattenRoute(nestedRoute);
  if (!routes || routes.length === 0) {
    return ``;
  }

  const clientLoaders: [string, string][] = [];
  for (const route of routes) {
    if (route[CLIENT_LOADER]) {
      clientLoaders.push([route.id, route.componentPath]);
    }
  }

  if (clientLoaders.length === 0) {
    return ``;
  }

  let imports = ``;
  let exports = `export default {\n`;

  for (let i = 0; i < clientLoaders.length; i++) {
    const [routeId, filename] = clientLoaders[i];
    imports += `import { ${CLIENT_LOADER} as loader_${i} } from '${filename}';\n`;
    exports += `
      ['${routeId}']: loader_${i},
    `;
  }

  exports += `\n}`;

  return `
    ${imports}
    ${exports}
  `;
};

export const fileSystemRoutes = ({
  routes,
  nestedRoute,
}: {
  routes?: Route[];
  nestedRoute?: NestedRoute;
}) => {
  const importLoadableCode = `import loadable from '@modern-js/runtime/loadable';`;
  let importLoadersCode = ``;
  let loadablePagesCode = '';
  let routesCode = '';
  let loadableNestedRoutesCode = '';
  let nestedRoutesCode = '';

  if (routes && routes.length > 0) {
    loadablePagesCode = routes
      .map(
        ({ component, _component }) =>
          `const ${component} = loadable(() => import('${_component}'));`,
      )
      .join('\n\n');

    routesCode = `
    export const routes = ${JSON.stringify(routes, null, 2).replace(
      /"component"\s*:\s*"(\S+)"/g,
      '"component": $1',
    )}
  ;`;
  }

  if (nestedRoute) {
    const routes = flattenRoute(nestedRoute);
    const useClientLoader = routes.some(route => route[CLIENT_LOADER]);
    if (useClientLoader) {
      importLoadersCode = `import _loaders from "./loaders";`;
    }

    loadableNestedRoutesCode = `
      export const routeComponents = {
    `;

    for (const route of routes) {
      const finalRoute: Partial<NestedRoute> & {
        loader: string;
        component: string;
      } = {
        ...route,
        loader: `_loaders['${route.id}']`,
        component: `loadable(() => import('${route.componentPath}'))`,
      };

      delete finalRoute.children;
      delete finalRoute?.componentPath;

      loadableNestedRoutesCode += `
        '${route.id}': ${JSON.stringify(finalRoute, null, 2)},
      `
        .replace(/"(loadable[^"]*)"/g, '$1')
        .replace(/"(_loaders[^"]*)"/g, '$1');
    }

    loadableNestedRoutesCode += `\n};`;

    nestedRoutesCode = `export const nestedRoutes = ${JSON.stringify(
      nestedRoute,
    )};`;
  }

  return `
    ${importLoadableCode}
    ${importLoadersCode}
    ${loadablePagesCode}
    ${routesCode}
    ${loadableNestedRoutesCode}
    ${nestedRoutesCode}
  `;
};
