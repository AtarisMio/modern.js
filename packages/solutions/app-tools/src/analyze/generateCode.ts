import path from 'path';
import { fs } from '@modern-js/utils';
import type {
  IAppContext,
  NormalizedConfig,
  PluginAPI,
  ImportSpecifier,
  ImportStatement,
} from '@modern-js/core';
import type { Entrypoint } from '@modern-js/types';
import esbuild from 'esbuild';
import * as templates from './templates';
import { getClientRoutes } from './getClientRoutes';
import {
  FILE_SYSTEM_ROUTES_FILE_NAME,
  ENTRY_POINT_FILE_NAME,
  ROUTES_DIR_NAME,
  LOADERS_FILE_TS,
  LOADERS_FILE_JS,
} from './constants';
import { getDefaultImports } from './utils';
import { getNestedRoutes } from './nestedRoutes';

const loader: { [ext: string]: esbuild.Loader } = {
  '.aac': 'file',
  '.css': 'text',
  '.less': 'text',
  '.sass': 'text',
  '.scss': 'text',
  '.eot': 'file',
  '.flac': 'file',
  '.gif': 'file',
  '.ico': 'file',
  '.jpeg': 'file',
  '.jpg': 'file',
  '.js': 'jsx',
  '.jsx': 'jsx',
  '.json': 'json',
  '.md': 'jsx',
  '.mdx': 'jsx',
  '.mp3': 'file',
  '.mp4': 'file',
  '.ogg': 'file',
  '.otf': 'file',
  '.png': 'file',
  '.svg': 'file',
  '.ts': 'ts',
  '.tsx': 'tsx',
  '.ttf': 'file',
  '.wav': 'file',
  '.webm': 'file',
  '.webp': 'file',
  '.woff': 'file',
  '.woff2': 'file',
};

export const EXTERNAL_REGEXP = /^[^./]|^\.[^./]|^\.\.[^/]/;

const createImportSpecifier = (specifiers: ImportSpecifier[]): string => {
  let defaults = '';

  const named = [];

  for (const { local, imported } of specifiers) {
    if (local && imported) {
      named.push(`${imported} as ${local}`);
    } else if (local) {
      defaults = local;
    } else {
      named.push(imported);
    }
  }

  if (defaults && named.length) {
    return `${defaults}, { ${named.join(', ')} }`;
  } else if (defaults) {
    return defaults;
  } else {
    return `{ ${named.join(', ')} }`;
  }
};

export const createImportStatements = (
  statements: ImportStatement[],
): string => {
  // merge import statements with the same value.
  const deDuplicated: ImportStatement[] = [];

  const seen = new Map();

  for (const { value, specifiers, initialize } of statements) {
    if (!seen.has(value)) {
      deDuplicated.push({
        value,
        specifiers,
        initialize,
      });
      seen.set(value, specifiers);
    } else {
      seen.get(value).push(...specifiers);
      // make "initialize" param can be connected when multiple plugins were imported from same package
      const modifyIndex = deDuplicated.findIndex(v => v.value === value);
      const originInitialize = deDuplicated[modifyIndex]?.initialize ?? '';
      deDuplicated[modifyIndex].initialize = originInitialize.concat(
        `\n${initialize || ''}`,
      );
    }
  }

  return deDuplicated
    .map(
      ({ value, specifiers, initialize }) =>
        `import ${createImportSpecifier(specifiers)} from '${value}';\n${
          initialize || ''
        }`,
    )
    .join('\n');
};

export const generateCode = async (
  appContext: IAppContext,
  config: NormalizedConfig,
  entrypoints: Entrypoint[],
  api: PluginAPI,
) => {
  const {
    internalDirectory,
    srcDirectory,
    internalDirAlias,
    internalSrcAlias,
  } = appContext;

  const hookRunners = api.useHookRunners();

  const {
    output: { mountId },
  } = config;

  for (const entrypoint of entrypoints) {
    const { entryName, isAutoMount, customBootstrap, fileSystemRoutes } =
      entrypoint;
    if (isAutoMount) {
      // TODO: 支持无 pages dir 的情况 @yimingjfe
      // generate routes file for file system routes entrypoint.
      if (fileSystemRoutes) {
        const initialRoutes = getClientRoutes({
          entrypoint,
          srcDirectory,
          srcAlias: internalSrcAlias,
          internalDirectory,
          internalDirAlias,
        });

        const routesDir = path.join(srcDirectory, ROUTES_DIR_NAME);
        const nestedRoute = await getNestedRoutes(routesDir);

        const { routes } = await hookRunners.modifyFileSystemRoutes({
          entrypoint,
          routes: initialRoutes,
        });

        const { code } = await hookRunners.beforeGenerateRoutes({
          entrypoint,
          code: templates.fileSystemRoutes({ routes, nestedRoute }),
        });

        if (nestedRoute) {
          const tmpLoadersFile = path.resolve(
            internalDirectory,
            `./${entryName}/${LOADERS_FILE_TS}`,
          );
          const outputLoadersFile = path.resolve(
            internalDirectory,
            `./${entryName}/${LOADERS_FILE_JS}`,
          );

          const clientLoaderCode = templates.getClientLoaderCode(nestedRoute);
          fs.outputFileSync(tmpLoadersFile, clientLoaderCode, 'utf8');

          await esbuild.build({
            format: 'esm',
            platform: 'browser',
            target: 'esnext',
            loader,
            watch: process.env.NODE_ENV === 'development' && {},
            bundle: true,
            logLevel: 'error',
            entryPoints: [tmpLoadersFile],
            outfile: outputLoadersFile,
            plugins: [
              {
                name: 'make-all-packages-external',
                setup(build) {
                  // https://github.com/evanw/esbuild/issues/619#issuecomment-751995294
                  build.onResolve({ filter: EXTERNAL_REGEXP }, args => {
                    let external = true;
                    // FIXME: windows external entrypoint
                    if (args.kind === 'entry-point') {
                      external = false;
                    }
                    return {
                      path: args.path,
                      external,
                    };
                  });
                },
              },
            ],
          });
        }

        fs.outputFileSync(
          path.resolve(
            internalDirectory,
            `./${entryName}/${FILE_SYSTEM_ROUTES_FILE_NAME}`,
          ),
          code,
          'utf8',
        );
      }

      // call modifyEntryImports hook
      const { imports: importStatements } =
        await hookRunners.modifyEntryImports({
          entrypoint,
          imports: getDefaultImports({
            entrypoint,
            srcDirectory,
            internalSrcAlias,
            internalDirAlias,
          }),
        });

      // call modifyEntryRuntimePlugins hook
      const { plugins } = await hookRunners.modifyEntryRuntimePlugins({
        entrypoint,
        plugins: [],
      });

      // call modifyEntryRenderFunction hook
      const { code: renderFunction } =
        await hookRunners.modifyEntryRenderFunction({
          entrypoint,
          code: templates.renderFunction({
            plugins,
            customBootstrap,
            fileSystemRoutes,
          }),
        });

      // call modifyEntryExport hook
      const { exportStatement } = await hookRunners.modifyEntryExport({
        entrypoint,
        exportStatement: 'export default AppWrapper;',
      });

      const code = templates.index({
        mountId: mountId as string,
        imports: createImportStatements(importStatements),
        renderFunction,
        exportStatement,
      });

      // generate entry file.
      const entryFile = path.resolve(
        internalDirectory,
        `./${entryName}/${ENTRY_POINT_FILE_NAME}`,
      );

      entrypoint.entry = entryFile;

      fs.outputFileSync(entryFile, code, 'utf8');
    }
  }
};
