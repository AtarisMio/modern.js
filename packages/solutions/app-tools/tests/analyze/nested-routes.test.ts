import path from 'path';
import { getNestedRoutes } from '../../src/analyze/nestedRoutes';
import { getClientLoaderCode } from '../../src/analyze/templates';

const fixtures = path.join(__dirname, 'fixtures');

describe('nested routes', () => {
  test('getNestedRoutes', async () => {
    const routeDir = path.join(fixtures, './nest-routes');
    const route = await getNestedRoutes(routeDir);
    expect(route).toMatchSnapshot();
  });

  test('getClientLoaderCode', async () => {
    const routeDir = path.join(fixtures, './nest-routes');
    const route = await getNestedRoutes(routeDir);
    const code = getClientLoaderCode(route!);
    expect(code).toMatchSnapshot();
  });
});
