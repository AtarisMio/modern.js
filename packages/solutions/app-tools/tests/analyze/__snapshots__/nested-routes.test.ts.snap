// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nested routes getClientLoaderCode 1`] = `""`;

exports[`nested routes getNestedRoutes 1`] = `
Object {
  "childIds": Array [
    "index",
    "sales",
  ],
  "children": Array [
    Object {
      "childIds": Array [],
      "children": Array [],
      "clientLoader": false,
      "componentPath": "/Users/bytedance/Desktop/workspaces/modern.js/packages/solutions/app-tools/tests/analyze/fixtures/nest-routes/index.tsx",
      "id": "index",
      "index": true,
      "path": "index",
      "root": false,
    },
    Object {
      "childIds": Array [
        "sales/invoices",
      ],
      "children": Array [
        Object {
          "childIds": Array [
            "sales/invoices/[id]",
            "sales/invoices/index",
          ],
          "children": Array [
            Object {
              "childIds": Array [],
              "children": Array [],
              "clientLoader": false,
              "componentPath": "/Users/bytedance/Desktop/workspaces/modern.js/packages/solutions/app-tools/tests/analyze/fixtures/nest-routes/sales/invoices/[id].tsx",
              "id": "sales/invoices/[id]",
              "index": false,
              "path": ":id",
              "root": false,
            },
            Object {
              "childIds": Array [],
              "children": Array [],
              "clientLoader": false,
              "componentPath": "/Users/bytedance/Desktop/workspaces/modern.js/packages/solutions/app-tools/tests/analyze/fixtures/nest-routes/sales/invoices/index.tsx",
              "id": "sales/invoices/index",
              "index": true,
              "path": "index",
              "root": false,
            },
          ],
          "clientLoader": false,
          "componentPath": "/Users/bytedance/Desktop/workspaces/modern.js/packages/solutions/app-tools/tests/analyze/fixtures/nest-routes/sales/invoices.tsx",
          "id": "sales/invoices",
          "index": false,
          "path": "invoices",
          "root": false,
        },
      ],
      "clientLoader": false,
      "componentPath": "/Users/bytedance/Desktop/workspaces/modern.js/packages/solutions/app-tools/tests/analyze/fixtures/nest-routes/sales.tsx",
      "id": "sales",
      "index": false,
      "path": "sales",
      "root": false,
    },
  ],
  "clientLoader": false,
  "componentPath": "/Users/bytedance/Desktop/workspaces/modern.js/packages/solutions/app-tools/tests/analyze/fixtures/nest-routes/app.tsx",
  "id": "app",
  "index": false,
  "path": "/",
  "root": true,
}
`;
